src/main/java/io/penguinstats/controller/v2/ReportControllerV2.java

```diff
@@ -0,0 +1,215 @@
+	public ResponseEntity<List<ItemDrop>> getPersonalReportHistory(HttpServletRequest request,
+			@RequestParam(name = "page", defaultValue = "0") Integer page,
+			@RequestParam(name = "page_size", defaultValue = "50") Integer pageSize,
+			@RequestParam(name = "sort_by", defaultValue = "timestamp") String sortBy,
+			@RequestParam(name = "direction", defaultValue = "ASC") String direction) {
+		try {
+			String userID = cookieUtil.readUserIDFromCookie(request);
+			if (userID == null) {
+				logger.error("Error in getPersonalReportHistory: Cannot read user ID");
+				return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+			}
+			logger.info("user " + userID + " GET /report/history\n");
+			Pageable pageable = PageRequest.of(page, pageSize, new Sort(Sort.Direction.fromString(direction), sortBy));
+			Page<ItemDrop> userItemDrops = itemDropService.getVisibleItemDropsByUserID(userID, pageable);
+			return new ResponseEntity<>(userItemDrops.getContent(), HttpStatus.OK);
+		} catch (Exception e) {
+			logger.error("Error in getPersonalReportHistory", e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	@PostMapping(path = "/history")
+	public ResponseEntity<String> deletePersonalReportHistory(HttpServletRequest request,
+			@RequestParam("item_drop_id") String itemDropId) {
+		try {
+			String userID = cookieUtil.readUserIDFromCookie(request);
+			if (userID == null) {
+				logger.error("Error in deletePersonalReportHistory: Cannot read user ID");
+				return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+			}
+
+			logger.info("user " + userID + " POST /report/history\n");
+			itemDropService.deleteItemDrop(userID, itemDropId);
+			return new ResponseEntity<>(HttpStatus.OK);
+		} catch (Exception e) {
+			logger.error("Error in deletePersonalReportHistory", e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	@ApiOperation("Recall the last report")
+	@PostMapping(path = "/recall")
+	public ResponseEntity<String> recallPersonalReport(HttpServletRequest request,
+			@RequestParam("item_drop_hash_id") String itemDropHashId) {
+		try {
+			String userID = cookieUtil.readUserIDFromCookie(request);
+			if (userID == null) {
+				logger.error("Error in recallPersonalReport: Cannot read user ID");
+				return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+			}
+
+			logger.info("user " + userID + " POST /report/recall\n");
+			itemDropService.recallItemDrop(userID, itemDropHashId);
+			return new ResponseEntity<>(HttpStatus.OK);
+		} catch (Exception e) {
+			logger.error("Error in recallPersonalReport", e);
+			return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+		}
+	}
+
+	private boolean isValidSingleReportRequest(String jsonString) {
+		try {
+			JSONObject obj = new JSONObject(jsonString);
+			if (!hasValidValue(obj, "stageId") || !hasValidValue(obj, "furnitureNum") || !hasValidValue(obj, "drops")) {
+				return false;
```

Comment:
why not just return `hasValidValue(obj, "stageId") && hasValidValue(obj, "furnitureNum") && hasValidValue(obj, "drops")`

Reply:
Yeah this is better... but the whole class needs to be refactored so no worry on this(x